docs/docs/server-middleware.md

```diff
@@ -0,0 +1,654 @@
+import org.http4s.server.middleware.HttpMethodOverrider.{HttpMethodOverriderConfig, QueryOverrideStrategy}
+
+val overrideService = HttpMethodOverrider(
+  service,
+  HttpMethodOverriderConfig(
+    QueryOverrideStrategy(paramName = "realMethod"),
+    Set(Method.GET)
+  )
+).orNotFound
+val overrideRequest = Request[IO](Method.GET, uri"/post?realMethod=POST")
+```
+```scala mdoc
+service.orNotFound(overrideRequest).unsafeRunSync().status
+overrideService(overrideRequest).unsafeRunSync().status
+```
+
+### HttpsRedirect
+Redirects requests to https when the `X-Forwarded-Proto` header is `http`. This
+header is usually provided by a load-balancer to indicate which protocol the client used.  
+
+```scala mdoc:silent
+import org.http4s.server.middleware.HttpsRedirect
+
+val httpsRedirectService = HttpsRedirect(service).orNotFound
+val httpRequest = okRequest
+  .putHeaders("Host" -> "example.com", "X-Forwarded-Proto" -> "http")
+```
+```scala mdoc
+httpsRedirectService(httpRequest).map(r => (r.headers, r.status)).unsafeRunSync()
+```
+
+### TranslateUri
+Removes a prefix from the path of the requested url.
+
+```scala mdoc:silent
+import org.http4s.server.middleware.TranslateUri
+
+val translateService = TranslateUri(prefix = "a")(service).orNotFound
+val translateRequest = Request[IO](Method.GET, uri"/a/b/c")
+```
+The following is successful even though `/b/c` is defined, and not `/a/b/c`:
+```scala mdoc
+translateService(translateRequest).unsafeRunSync().status
+```
+
+### UrlFormLifter
+
+Transform `x-www-form-urlencoded` parameters into query parameters.
+
+```scala mdoc:silent
+import org.http4s.server.middleware.UrlFormLifter
+import org.http4s.UrlForm
+
+val urlFormService = UrlFormLifter.httpApp(service.orNotFound)
+val formRequest = Request[IO](Method.POST, uri"/queryForm")
+  .withEntity(UrlForm.single("name", "John"))
+```
+Even though the `/queryForm` route takes query parameters, the form request works:
+```scala mdoc
+urlFormService(formRequest).flatMap(_.bodyText.compile.lastOrError).unsafeRunSync()
+```
+
+## Scaling and resource management
+
+### ConcurrentRequests
+
+React to requests being accepted and completed, could be used for metrics.
+
+```scala mdoc:silent
+import org.http4s.server.middleware.ConcurrentRequests
+import org.http4s.server.{ContextMiddleware, HttpMiddleware}
+import org.http4s.ContextRequest
+import cats.data.Kleisli
+import cats.effect.Ref
+
+// a utility that drops the context from the request, since our service expects
+// a plain request
+def dropContext[A](middleware: ContextMiddleware[IO, A]): HttpMiddleware[IO] =
+  routes => middleware(Kleisli((c: ContextRequest[IO, A]) => routes(c.req)))
+
+val concurrentLog = Ref[IO].of(List.empty[String]).unsafeRunSync()
+
+val concurrentService =
+  ConcurrentRequests.route[IO](
+      onIncrement = total => 
+        concurrentLog.update(_ :+ s"someone comes to town, total=$total"),
+      onDecrement = total => 
```

Comment:
I wrongly thought that printlns wouldn't be rendered by mdoc, but it turns out that it's just `IO.println` that doesn't work `IO(println(...))` does

Reply:
Hmm, not crazy about the `IO(println(...))`, since that's an anti-pattern (did you hear Daniel's story about how this has melt down some systems in production 😛 ). What do you think about defining our own implicit `Console[IO]` in a hidden code block that delegates to `IO(println(...))`? Then we can do `Console[IO].println(...)`. Yes I know a lot of boilerplate 😕

Reply:
Didn't think about that, i'll do it

Reply:
I'm having a time
```scala
import cats.Show

implicit val mdocConsoleIO: Console[IO] = new Console[IO] {
  def println[A](a: A)(implicit s: Show[A] = Show.fromToString[A]): IO[Unit] = {
    val str = s.show(a)
    IO.blocking(scala.Console.println(str)) >> 
      IO.blocking(println(str)) >>
      IO.blocking(System.out.println(str)) >>
      IO(scala.Console.println(str)) >> 
      IO(println(str)) >>
      IO(System.out.println(str))
  }

  def print[A](a: A)(implicit S: Show[A] = Show.fromToString[A]): IO[Unit] = IO.unit
  def error[A](a: A)(implicit S: Show[A] = Show.fromToString[A]): IO[Unit] = IO.unit
  def errorln[A](a: A)(implicit S: Show[A] = Show.fromToString[A]): IO[Unit] = IO.unit
  def readLineWithCharset(charset: java.nio.charset.Charset): IO[String] = IO.pure("")
}
// mdocConsoleIO: Console[IO] = repl.MdocSession$App$$anon$1@56428a69

// checking that the summoned Console[IO] is this one ☝️
Console[IO] 
// res0: Console[IO] = repl.MdocSession$App$$anon$1@56428a69 

(
  Console[IO].println("1") >>
    mdocConsoleIO.println("2") >>
    IO.blocking(println("3"))
).unsafeRunSync()
// 3
```

Reply:
This is relevant
https://github.com/scalameta/mdoc/issues/517

Rob has a slightly different usage (the output is rendered directly instead of being a comment below the expression):
https://tpolecat.github.io/skunk/tutorial/Query.html#full-example

https://github.com/tpolecat/skunk/blob/20126c123a2eb3a191d90b8fa59c4ead3391e6ff/modules/docs/src/main/paradox/tutorial/Query.md?plain=1#L314-L321

Reply:
@armanbilge Rob's unsafeRunSyncWithRedirect() works:

```scala
implicit class IOOps1[A](fa: IO[A]) {
  def unsafeRunSync(): A = {
    val oldOut = System.out
    val newOut = scala.Console.out // important to do this on the calling thread!
    try {
      System.setOut(newOut)
      fa.unsafeRunSync()
    } finally System.setOut(oldOut)
  }
}



new IOOps1(
  Console[IO].println("1") >>
    mdocConsoleIO.println("2") >>
    IO.blocking(println("3"))
  ).unsafeRunSync()
// 1
// 1
// 2
// 2
// 3
```

can I trick the compiler into picking up this `unsafeRunSync`?

Reply:
Hm, I'm confused what's going on 😂 thanks for trying all that. Ok, how about this: can we define our `Console[IO]#println` something like this?

```scala
IO.bracket {
  IO.blocking {
    val oldOut = System.out
    val newOut = scala.Console.out
    System.setOut(newOut)
    oldOut
  }
}(_ => IO.println(...))(oldOut => IO(System.setOut(oldOut)))
```

Reply:
Oh wait 🤔 
```scala
val newOut = Console.out // important to do this on the calling thread!
```

Ok, scratch the above.

What about this.

```scala
val mdocConsoleOut = Console.out // important to do this on the calling thread!
implicit val mdocConsoleIO: Console[IO] = new Console[IO] {
  def println[A](a: A)(implicit s: Show[A] = Show.fromToString[A]): IO[Unit] = {
    val str = s.show(a)
    IO(mdocConsoleOut.println(str)) >>
  }
```

Reply:
That works!

Reply:
I've pushed this change